File: bot/src/main/java/by/yayauheny/tutochkatgbot/dto/backend/RestroomResponseDto.java
package by.yayauheny.tutochkatgbot.dto.backend;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.time.Instant;
import java.util.Map;
import java.util.UUID;

/**
 * Restroom response DTO.
 * feeType, genderType, accessibilityType, name, address may be null (backend stores them as optional).
 */
public record RestroomResponseDto(
    @JsonProperty("id") UUID id,
    @JsonProperty("cityId") UUID cityId,
    @JsonProperty("buildingId") UUID buildingId,
    @JsonProperty("subwayStationId") UUID subwayStationId,
    @JsonProperty("name") String name,
    @JsonProperty("address") String address,
    @JsonProperty("phones") Map<String, Object> phones,
    @JsonProperty("workTime") Map<String, Object> workTime,
    @JsonProperty("feeType") FeeType feeType,
    @JsonProperty("genderType") GenderType genderType,
    @JsonProperty("accessibilityType") AccessibilityType accessibilityType,
    @JsonProperty("placeType") PlaceType placeType,
    @JsonProperty("coordinates") LatLon coordinates,
    @JsonProperty("dataSource") DataSourceType dataSource,
    @JsonProperty("status") RestroomStatus status,
    @JsonProperty("amenities") Map<String, Object> amenities,
    @JsonProperty("externalMaps") Map<String, Object> externalMaps,
    @JsonProperty("accessNote") String accessNote,
    @JsonProperty("directionGuide") String directionGuide,
    @JsonProperty("inheritBuildingSchedule") Boolean inheritBuildingSchedule,
    @JsonProperty("hasPhotos") Boolean hasPhotos,
    @JsonProperty("locationType") LocationType locationType,
    @JsonProperty("originProvider") ImportProvider originProvider,
    @JsonProperty("originId") String originId,
    @JsonProperty("isHidden") Boolean isHidden,
    @JsonProperty("createdAt") Instant createdAt,
    @JsonProperty("updatedAt") Instant updatedAt,
    @JsonProperty("distanceMeters") Integer distanceMeters,
    @JsonProperty("building") BuildingResponseDto building,
    @JsonProperty("subwayStation") SubwayStationResponseDto subwayStation
) {}

File: backend/src/main/kotlin/yayauheny/by/model/dto/RestroomResponseDto.kt
package yayauheny.by.model.dto

import java.time.Instant
import java.util.UUID
import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import yayauheny.by.model.building.BuildingResponseDto
import yayauheny.by.model.enums.AccessibilityType
import yayauheny.by.model.enums.DataSourceType
import yayauheny.by.model.enums.FeeType
import yayauheny.by.model.enums.GenderType
import yayauheny.by.model.enums.LocationType
import yayauheny.by.model.enums.ImportProvider
import yayauheny.by.model.enums.PlaceType
import yayauheny.by.model.enums.RestroomStatus

@Serializable
data class RestroomResponseDto(
    @Contextual val id: UUID,
    @Contextual val cityId: UUID?,
    @Contextual val buildingId: UUID?,
    @Contextual val subwayStationId: UUID?,
    val name: String?,
    val address: String?,
    val phones: JsonObject?,
    val workTime: JsonObject?,
    val feeType: FeeType?,
    val genderType: GenderType?,
    val accessibilityType: AccessibilityType?,
    val placeType: PlaceType?,
    val coordinates: Coordinates,
    val dataSource: DataSourceType,
    val status: RestroomStatus,
    val amenities: JsonObject?,
    val externalMaps: JsonObject?,
    val accessNote: String?,
    val directionGuide: String?,
    val inheritBuildingSchedule: Boolean,
    val hasPhotos: Boolean,
    val locationType: LocationType,
    val originProvider: ImportProvider,
    val originId: String?,
    val isHidden: Boolean,
    @Contextual val createdAt: Instant,
    @Contextual val updatedAt: Instant,
    val distanceMeters: Int? = null,
    val building: BuildingResponseDto? = null,
    val subwayStation: SubwayStationResponseDto? = null
)
